modules/nlu/src/backend/engine2.ts

```diff
@@ -0,0 +1,1184 @@
+    const charset_low_score = _.intersection(charsetLow1, charsetLow2).length / _.union(charsetLow1, charsetLow2).length
+    const final_charset_score = _.mean([charset_score, charset_low_score])
+
+    const la = Math.max(1, a.filter(x => x.length > 1).length)
+    const lb = Math.max(1, a.filter(x => x.length > 1).length)
+    const token_qty_score = Math.min(la, lb) / Math.max(la, lb)
+
+    const size1 = _.sumBy(a, 'length')
+    const size2 = _.sumBy(b, 'length')
+    const token_size_score = Math.min(size1, size2) / Math.max(size1, size2)
+
+    return Math.sqrt(final_charset_score * token_qty_score * token_size_score)
+  }
+
+  const matches: EntityExtractionResult[] = []
+
+  for (const list of list_entities) {
+    const candidates = []
+    let longestCandidate = 0
+
+    for (const [canonical, occurances] of _.toPairs(list.mappingsTokens)) {
+      for (const occurance of occurances) {
+        for (let i = 0; i < utterance.tokens.length; i++) {
+          if (utterance.tokens[i].isSpace) {
+            continue
+          }
+          const workset = takeUntil(utterance.tokens, i, _.sumBy(occurance, 'length'))
+          const worksetAsStrings = workset.map(x => x.toString({ lowerCase: true, realSpaces: true, trim: false }))
+          const candidateAsString = occurance.join('')
+
+          if (candidateAsString.length > longestCandidate) {
+            longestCandidate = candidateAsString.length
+          }
+
+          const fuzzy = list.fuzzyMatching && worksetAsStrings.join('').length >= 4
+          const exact_score = exactScore(worksetAsStrings, occurance) === 1 ? 1 : 0
+          const fuzzy_score = fuzzyScore(worksetAsStrings, occurance)
+          const structural_score = structuralScore(worksetAsStrings, occurance)
+          const finalScore = fuzzy ? fuzzy_score * structural_score : exact_score * structural_score
+
+          candidates.push({
+            score: Math.round(finalScore * 1000) / 1000,
+            canonical,
+            start: i,
+            end: i + workset.length - 1,
+            source: workset.map(t => t.toString({ lowerCase: false, realSpaces: true })).join(''),
+            occurance: occurance.join(''),
+            eliminated: false
+          })
+        }
+      }
+
+      for (let i = 0; i < utterance.tokens.length; i++) {
+        const results = _.orderBy(
+          candidates.filter(x => !x.eliminated && x.start <= i && x.end >= i),
+          // we want to favor longer matches (but is obviously less important than score)
+          // so we take its length into account (up to the longest candidate)
+          x => x.score * Math.pow(Math.min(x.source.length, longestCandidate), 1 / 5),
+          'desc'
+        )
+        if (results.length > 1) {
+          const [, ...losers] = results
+          losers.forEach(x => (x.eliminated = true))
+        }
+      }
+    }
+
+    candidates
+      .filter(x => !x.eliminated && x.score >= 0.65)
```

Comment:
Magic number that seems important should be a constant

Reply:
yes but that will be configurable soon. Fuzziness factor.